pkgs/jni/src/dartjni.c

```diff
@@ -146,6 +150,11 @@ BOOL WINAPI DllMain(HINSTANCE hinstDLL,   // handle to DLL module
       // Return FALSE to fail DLL load.
```

Comment:
sidenote: This `DllMain` is assuming the C code in `package:jni` gets compiled into it's own DLL (?). That would prevent statically linking it into a bigger library (which may have already a different `DllMain`). It's a little bit of a smell I think.

If we can use C++ here (if not: why?) then a global C++ object can do things in it's constructor and destructor (e.g. initializer locks and free them, create TLS keys, free them, ...) that get run on shared library load & unload times.

(Arguably this doesn't solve the issue of detaching the thread from java at exit)

Reply:
We could use C++. We would have to wrap the functions we want to call from Dart in `extern "C"` but otherwise it should not be a problem. We also generate another C file from `jni.h` which wraps the API and does things like conversion to global refs and exception checking. 

In general, I did not really refactor the C code we had since 2 years ago. I think it's worth it to open an issue and do this in the near future. https://github.com/dart-lang/native/issues/1169